doc/api/esm.md

```diff
@@ -596,26 +600,33 @@ CommonJS modules loaded.
 
 #### `resolve(specifier, context, defaultResolve)`
 
+<i id="esm_resolve_hook"></i>
+
 > Note: The loaders API is being redesigned. This hook may disappear or its
 > signature may change. Do not rely on the API described below.
 
 * `specifier` {string}
 * `context` {Object}
   * `conditions` {string[]}
-  * `parentURL` {string}
-* `defaultResolve` {Function}
+  * `parentURL` {string?}
+* `defaultResolve` {Function} The Node.js default resolver.
 * Returns: {Object}
-  * `url` {string}
+  * `format` {string?} Any value is allowed, but ensure `load` hook's final
+  value is one Node.js understands.
+  * `url` {string} The fully qualified url to the import target (`file://…`,
+  `http://…`, etc)
 
 The `resolve` hook returns the resolved file URL for a given module specifier
-and parent URL. The module specifier is the string in an `import` statement or
-`import()` expression, and the parent URL is the URL of the module that imported
-this one, or `undefined` if this is the main entry point for the application.
+and parent URL, and optionally its format (ex `'module'`); falsy format values
+are ignored. The module specifier is the string in an `import` statement or
```

Comment:
I opted for falsy because anything falsy is guaranteed ingore-able:

* `undefined`: invalid & useless
* `null`: invalid & useless
* `false`: invalid & useless
* `0` (zero): invalid & useless
* `''` (empty string): valid, but still useless

I'm not married to falsy over nullish, but nullish seems like splitting hairs in a way to facilitate the user accidentally screwing themselves in a way that's easily available. Happy to be shown wrong though.

Reply:
I'd argue that using nullish over falsy would help users avoid screwing themselves 😅 If an hypothetical user mistakenly uses `process.env.NODE_ENV === 'production'` as value for the `format`, they won't know until the code hits production that there's a mistake. If we check for nullish value, it would send an helpful error, allowing them to spot the error quicker.

Reply:
If I'm understanding your hypothetical correctly (and ignoring that hooks should not be used in production 😜), I believe that is not true: the `format` returned by `resolve()` can be anything, **but** it must become one of the 5 values (builtin, commonjs, module, json, wasm) Node.js understands when `load()` eventually returns the final/official value. So if a user did do:

```js
async function resolve(…) {
  return {
    format: process.env.NODE_ENV, // 'production'
    url: 'whatever',
  };
}
```
```js
async function load(url, { format }, defaultLoader) {
  return {
    format, // 'production'
    source: 'export default "i explode"',
  };
}
```

That would fail, telling the user the format value returned by the load hook was invalid.

Or are you saying a user sets `NODE_ENV` only in production? If so, they're boned regardless of falsy vs nullish.

Reply:
It was a silly example I reckon:

```js
async function resolve(…) {
  return {
    format: process.env.NODE_ENV === 'production'
    url: 'whatever',
  };
}
```

During development, `format` is falsy (`false`) and when pushed to production it becomes truthy (`true`). Almost certainly the developer didn't mean to write this code, if Node.js checks for nullish values instead of falsy, an error is thrown during development helping them to spot it before it hits production.

Reply:
Ahhh. Gotcha. I'm convinced 🙂 will update either this evening or tomorrow (getting vaccinated today, so might be outta commission tonight).